internal/backend/local/test.go

```diff
@@ -1262,22 +1262,75 @@ func (runner *TestFileRunner) prepareInputVariablesForAssertions(config *configs
 func (runner *TestFileRunner) ctx(run *moduletest.Run, file *moduletest.File, availableVariables terraform.InputValues) (*hcl.EvalContext, tfdiags.Diagnostics) {
 	var diags tfdiags.Diagnostics
 
-	availableRunBlocks := make(map[string]bool)
+	// First, let's build the set of available run blocks.
+
+	availableRunBlocks := make(map[string]*terraform.TestContext)
+	runs := make(map[string]cty.Value)
 	for _, run := range file.Runs {
 		name := run.Name
 
-		if _, exists := runner.PriorStates[name]; exists {
+		attrs := make(map[string]cty.Value)
+		if ctx, exists := runner.PriorStates[name]; exists {
 			// We have executed this run block previously, therefore it is
 			// available as a reference at this point in time.
-			availableRunBlocks[name] = true
+			availableRunBlocks[name] = ctx
+
+			for name, config := range ctx.Config.Module.Outputs {
+				output := ctx.State.OutputValue(addrs.AbsOutputValue{
+					OutputValue: addrs.OutputValue{
+						Name: name,
+					},
+					Module: addrs.RootModuleInstance,
+				})
+
+				var value cty.Value
+				switch {
+				case output == nil:
+					// This means the run block returned null for this output.
+					// It is likely this will produce an error later if it is
+					// referenced, but users can actually specify that null
+					// is an acceptable value for an input variable so we won't
+					// actually raise a fuss about this at all.
+					value = cty.NullVal(cty.DynamicPseudoType)
+				case output.Value.IsNull() || output.Value == cty.NilVal:
+					// This means the output value was returned as (known after
```

Comment:
I don't think this is a behavior you can always count on for that situation. There's nothing to guarantee that an unknown output should map to a literal `null` in the state (and nothing should be saving a `cty.NilVal`), so this is something that is prone to breaking in the future if the handling of null values changes at all. (I'm particularly thinking of other projects' influence where more fidelity may need to be recorded between runs)

I don't know if there's anything better we can do at the moment, so maybe just a full test of the situation to catch breakage is sufficient for now.

Reply:
The bit that I saw that made me think the behaviour is okay is here: https://github.com/hashicorp/terraform/blob/main/internal/terraform/node_output.go#L604

Basically, if something is null its not included in the output at all, and then unknown root module outputs are marked as null. At this point, we're only processing root module outputs so that was my thinking here.

We are testing this end to end within the tests I have, so they should start failing if this assumption changes.